File: android/app/src/main/kotlin/com/example/test_twilio/arguments/MessageItemArgument.kt
package com.example.test_twilio.arguments

import android.util.Log
import com.example.test_twilio.ChatCallback
import com.twilio.conversations.Attributes
import com.twilio.conversations.Message
import com.twilio.conversations.Participant

class MessageItemArgument(val message: Message, val members: List<Participant>) {

    companion object {
        fun toMapList(list: MutableList<MessageItemArgument>, listener: ChatCallback): List<HashMap<String, Any?>> {
            val listHashMap = mutableListOf<HashMap<String, Any?>>()
            list.let {
                for (messageItem in it) {
                    listHashMap.add(messageItem.toMap(listener))
                }
            }
            return listHashMap
        }
    }
    fun toMap(listener: ChatCallback?): HashMap<String, Any?> {
        val hashMap = hashMapOf<String, Any?>()
        hashMap["message"] = messageMap(message, listener)
        hashMap["members"] = membersMap(members)
        return hashMap
    }

    private fun messageMap(message: Message, listener: ChatCallback?) : HashMap<String, Any?> {
        val hashMap = hashMapOf<String, Any?>()
        hashMap["sid"] = message.sid
        hashMap["author"] = message.author
        hashMap["dateCreated"] = message.dateCreated
        hashMap["dateUpdated"] = message.dateUpdated
        hashMap["lastUpdatedBy"] = message.lastUpdatedBy
        hashMap["messageBody"] = message.messageBody
        hashMap["conversationSid"] = message.conversationSid
        hashMap["participantSid"] = message.participantSid
        hashMap["messageIndex"] = message.messageIndex
        hashMap["type"] = message.type.value
        hashMap["hasMedia"] = message.hasMedia()
        if (message.hasMedia()) {
            hashMap["mediaFileName"] = message.mediaFileName
            hashMap["mediaSid"] = message.mediaSid
            hashMap["mediaType"] = message.mediaType
            hashMap["mediaSize"] = message.mediaSize
            message.getMediaContentTemporaryUrl {
                Log.e(this@MessageItemArgument.javaClass.simpleName, it)
                listener?.getMediaPathComplete(MediaFilePathArgument(it, message.sid))
            }
        }
        hashMap["attributes"] = message.attributes.toString()
        return hashMap
    }

    private fun membersMap(members: List<Participant>) : HashMap<String, Any?> {
        val hashMap = hashMapOf<String, Any?>()
        hashMap["members"] = memberMap(members)
        return hashMap
    }

    private fun memberMap(membersList: List<Participant>) : List<HashMap<String, Any?>> {
        val listHashMap = mutableListOf<HashMap<String, Any?>>()
        for (member in membersList) {
            val hashMap = hashMapOf<String, Any?>()
            hashMap["sid"] = member.sid
            hashMap["lastReadMessageIndex"] = member.lastReadMessageIndex
            hashMap["lastReadTimestamp"] = member.lastReadTimestamp
            hashMap["identity"] = member.identity
            hashMap["type"] = member.type.value
            listHashMap.add(hashMap)
        }
        return listHashMap
    }
}

File: ios/Runner/arguments/MessageItemArgument.swift
//
//  MessageItemArgument.swift
//  Runner
//
//  Created by KienNT on 10/26/21.
//

import Foundation
import TwilioConversationsClient

class MessageItemArgument {
    var message: TCHMessage?
    var members: [TCHParticipant]?
    
    init(message: TCHMessage?, members: [TCHParticipant]?) {
        self.message = message
        self.members = members
    }
    
    static func toMapList(list: [MessageItemArgument]) -> [[String: Any?]] {
        var listMap: [[String: Any?]] = []
        for messageItem in list {
            listMap.append(messageItem.toMap())
        }
        return listMap
    }
    
    func toMap() -> [String: Any?] {
        var map: [String: Any?] = [:]
        map["message"] = messageMap(message: message!)
        map["members"] = membersMap(members: members!)
        return map
    }
    
    private func messageMap(message: TCHMessage) -> [String: Any?] {
        var map: [String: Any?] = [:]
        map["sid"] = message.sid
        map["author"] = message.author
        map["dateCreated"] = message.dateCreated
        map["dateUpdated"] = message.dateUpdated
        map["lastUpdatedBy"] = message.lastUpdatedBy
        map["messageBody"] = message.body
        map["participantSid"] = message.participantSid
        map["messageIndex"] = message.index
        map["type"] = message.messageType.rawValue
        map["hasMedia"] = message.hasMedia()
        return map
    }
    
    private func membersMap(members: [TCHParticipant]?) -> [String: Any?] {
        var map: [String: Any?] = [:]
        map["members"] = memberMap(membersList: members!)
        return map
    }
    
    private func memberMap(membersList: [TCHParticipant]) -> [[String: Any?]] {
        var listMap: [[String: Any?]] = []
        for member in membersList {
            var map: [String: Any?] = [:]
            map["sid"] = member.sid
            map["lastReadMessageIndex"] = member.lastReadMessageIndex
            map["lastReadTimestamp"] = member.lastReadTimestamp
            map["identity"] = member.identity
            map["type"] = member.type.rawValue
            listMap.append(map)
        }
        return listMap
    }
}
